observability/logging.go

```diff
@@ -0,0 +1,222 @@
+			grpcLogRecord.Authority = binlogEntry.GetClientHeader().Authority
+			grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetClientHeader().Metadata)
+		}
+		if binlogEntry.GetPeer() != nil {
+			grpcLogRecord.PeerAddress = &grpclogrecordpb.GrpcLogRecord_Address{
+				Type:    grpclogrecordpb.GrpcLogRecord_Address_Type(binlogEntry.Peer.Type),
+				Address: binlogEntry.Peer.Address,
+				IpPort:  binlogEntry.Peer.IpPort,
+			}
+		}
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_HEADER:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_RESPONSE_HEADER
+		if binlogEntry.Peer != nil {
+			grpcLogRecord.PeerAddress = &grpclogrecordpb.GrpcLogRecord_Address{
+				Type:    grpclogrecordpb.GrpcLogRecord_Address_Type(binlogEntry.Peer.Type),
+				Address: binlogEntry.Peer.Address,
+				IpPort:  binlogEntry.Peer.IpPort,
+			}
+		}
+		grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetServerHeader().Metadata)
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CLIENT_MESSAGE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_REQUEST_MESSAGE
+		grpcLogRecord.Message = binlogEntry.GetMessage().Data
+		grpcLogRecord.PayloadSize = binlogEntry.GetMessage().GetLength()
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_MESSAGE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_RESPONSE_MESSAGE
+		grpcLogRecord.Message = binlogEntry.GetMessage().Data
+		grpcLogRecord.PayloadSize = binlogEntry.GetMessage().GetLength()
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CLIENT_HALF_CLOSE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_HALF_CLOSE
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_TRAILER:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_TRAILER
+		grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetTrailer().Metadata)
+		grpcLogRecord.StatusCode = binlogEntry.GetTrailer().StatusCode
+		grpcLogRecord.StatusMessage = binlogEntry.GetTrailer().StatusMessage
+		grpcLogRecord.StatusDetails = binlogEntry.GetTrailer().StatusDetails
+		grpcLogRecord.PayloadTruncated = binlogEntry.PayloadTruncated
+		callEnded = true
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CANCEL:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_CANCEL
+		callEnded = true
+	default:
+		return fmt.Errorf("unknown event type: %v", binlogEntry.Type)
+	}
+	switch binlogEntry.Logger {
+	case binlogpb.GrpcLogEntry_LOGGER_CLIENT:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_CLIENT
+	case binlogpb.GrpcLogEntry_LOGGER_SERVER:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_SERVER
+	default:
+		grpcLogRecord.EventLogger = grpclogrecordpb.GrpcLogRecord_LOGGER_UNKNOWN
+	}
+	if callEnded {
```

Comment:
This depends on the `binarylog`'s implementation. I thought about several cases:

- Normal path, it's definitely cleaned;
- For errors returned by server, cleaned;
- For errors generated by the client (like DEADLINE_EXCEEDED), the event type will be `GrpcLogEntry_EVENT_TYPE_SERVER_TRAILER`, even though it's not sent by the server, cleaned;
- For cancellations, we only log client-side cancel, only 1 of cancel and trailer will be logged (see [code](https://github.com/grpc/grpc-go/blob/61a6a06b8879354998373f8bef56312ea07d6719/stream.go#L905)), it should be safe.

Do you have other cases?

Reply:
There are other possibilities.  What if the connection is lost?  We need to be 100% sure this will always be called, or we will want to find another way.  (E.g. can we attach a UUID for each RPC into the context via an interceptor/stats handler?  And will stats/tracing need this anyway?)